### PR TITLE
Added JSON logging, Docker improvements, client-side git hooks.

### PR DESCRIPTION
**Pull request by lyndonhenry**
## /src/*.{cpp,h}
- modified logging functionality to produce json output if -p/--profile option is given as a file with a .json extension
- refactored logging, statements now generated in src/AstLogStatement.h
- removed print json method in /src/PrecedenceGraph.{h,cpp} due to deprecation of -s
- add license header to src/Macro.h

## /tests/example/profile_{default,json}/*
- added example to show logging working with json extension
- also added example to show logging fallback to default on unrecognized extension

## /git/hooks, /git/hooks/pre-commit
- new git hooks directory for useful client side git hooks
- pre-commit hook to run clang format before each commit
- to use, just copy to your soffle/.git/hooks directory 

## /docker/Dockerfile.*, /docker/run.sh, /docker/souffle-docker.sh
- Dockerfiles and souffle-docker.sh have more informative variable names
- typos in Dockerfiles are fixed
- Dockerfiles use common run.sh script to avoid code duplication
- more friendly error handling in souffle-docker.sh

### CODE REVIEW

**Comment by mmcgr** [comment]
ok to test

**Comment by mmcgr** [comment]
That was to my bot, to test triggering jenkins testing, which should have automically triggered anyway.

**Comment by b-scholz** [comment]
This is only a general comment and not specific to this pull-request. Perhaps we go ahead with this pull-request and address the two challenges mentioned below in a separate issue.

First, we should have only one profile format. Hence, the reader of the profiler should be converted to the JSON format as well and the old file format should be retired. The main idea behind a single profiling format is to keep the usability of Soufflé simple. In case, somebody still wants to use old profiles we may need a conversion tools that converts the old profiles to new profiles in the new format. Although the conversion tool may be of less importance (i.e. an old version of souffle-profile still can be used).

Another concern with the current profiling tool is that it creates a lot of space overheads. It is quite verbose and hence it slows the profiling down at runtime. We may want consider improving the current profiling system by storing all events in memory and only at the end of execution, all events are written to disk. This will give us less distorted numbers. For this purpose we would need new data-structures capturing various space/time events at runtime. A printing facility will produce the profile in JSON format at the end of execution.

**Comment by mmcgr** [comment]
I think we're reaching the point where we need a utilities directory, into which the docker stuff and git hook examples would go. (Which should then really be in a separate souffle-utils repo, but that doesn't exist yet.) That would also be a good place for a script to turn json profiling into non-json. Docker scripts specifically for testing on travis have particular needs, or would if they existed, so would go into .travis.

For now, I think we should just move the docker/git stuff into /utilities/.

**Comment by lyndonhenry** [comment]
I like all of these ideas.

The reasoning for having one profiler format makes perfect sense. JSON seems like it should be a good choice for this, as it is simple and universal. If anything more standardized exists for this sort of task, that would be even more appropriate. I think if people want other formats, that's outside of the scope of Souffle, and they can write their own converter. Doing this will mean we obviously have to re-write some the profiler itself (i.e. not the logging functionality within souffle, but the tool that analyses these logs). This should be pretty trivial, but is probably best left to another PR. So I think perhaps an issue should be created for this.

Storing everything in memory and dumping at the end is substantially better than the current method, so I think another issue should be created for moving towards this also. 

Having separate repositories for the docker/git stuff gives us good separation of concerns, and should definitely be done. I think it would be best to have one repository for docker images (as we already have one at https://github.com/souffle-lang/souffle-docker), and then to have another for utilities (the proposed souffle-utils) where we put development tools, helper scripts, and other miscellaneous things which don't fit elsewhere but are not substantial enough to warrant their own repo. I can move the docker stuff over immediately, and also make a souffle-utils repo which souffle-lang can fork, or can move the docker and git hooks to a utilities directory and do everything with the other repos as a separate step. The former seems the preferable option, as it solves the problem immediately and with the most minimum effort. 

So, in summary:
- We move the profiler format to something standard, i.e. JSON, and then modify the profiler tool to use this new format. Maybe we also make a conversion utility to go from the new format to the old. **Should I create an issue for this?**
- We improve the current profiler to make it more efficient with output written only at the end of execution. **Should I create another issue for this?**
- We either 1) move the docker/git stuff to a /utilities directory, or, 2) move the docker stuff to souffle-docker and the git stuff to a new souffle-utils repo.  **Which of these is best?**

**Comment by mmcgr** [comment]
souffle-lang/souffle-utilities exists now. I think an improved profiler tool, and profiling output converter tools, should go there. The git stuff too, of course.

Docker stuff is less clear. I think they should be in utilities, rather than a separate docker repo. Is there an advantage to the separate souffle-docker repo? I'm new to the ways of docker, but would it be better if the images were in a dockerhub account for souffle-lang?

For now, go ahead with the optional json profiler output, remove the git stuff and make a PR to souffle-utilities, remove the docker stuff and send that to souffle-docker, and then make an issue for the further profiler change and an issue for the docker stuff.

File: src/AstLogStatement.h [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        if (AstLogStatement::getLogFileType() == "json") {
+            line << "{";
+            line << "\"msg-type\": \"" << messageType << "\", ";
+            line << "\"relation\": \"" << relationName << "\", ";
+            line << "\"src-locator\": \"" << srcLocation << "\", ";
+            line << "\"rule\": \"" << datalogText << "\", ";
+            line << "\"failed-proofs\": ";
+        } else {
+            line << messageType << ";" << relationName << ";" << srcLocation << ";" << datalogText << ";";
+            // TODO: the additional semicolon is added to maintain backwards compatibility and should
```
Comment (mmcgr) [inline src/AstLogStatement.h]: TODOs should be in the form
`// TODO (username): description`
or
`// TODO (#issue number): description`
In this case, the issue number for the new profiler issue.

**Comment by lyndonhenry** [comment]
All sounds very good, I'll do that.

The images should be on dockerhub, but the Dockerfiles themselves ought to have a git repository of their own, as they share common code. I'll attempt to set it up in souffle-docker that they may be built and uploaded to dockerhub in a souffle account, then transfer ownership of that account to an admin.

So in terms of the docker vs utilities stuff, we know the argument for having them in *either* souffle-utilities or souffle-docker and not in souffle is in that if both souffle and souffle-docker are in the same repository, all the dependencies of the souffle-docker become dependencies of souffle, and also all the dependencies of souffle become the dependencies of souffle-docker. The dependencies that they share are essentially parts of a system that can be relatively safely assumed (e.g. bash); they are entirely disjoint pieces of software.

The argument for having both souffle-docker and souffle-utilities as separate repositories is that apart from those assumed peices of software, souffle-docker has no other dependency but docker itself. This is very much "in the philosophy" of docker, as docker in part exists to eliminate the need to install dependencies.

That said, the two repository approach might be overengineering simplicity, and putting the souffle-docker code in the souffle-utilities repository might be easier to navigate, use, and develop. It's really a case of sphagetti code vs separation of concerns. I don't really mind where it goes to be honest, but I think it is important to have justification for why we are doing things in the way we are.

**Comment by lyndonhenry** [comment]
Travis seems to be having problems getting stuff from Launchpad, that is the build is failing for reasons unrelated to my PR. I'll try again later or, if that doesn't work, attempt to fix it tomorrow if someone else doesn't manage to before me.
